Support Landlock syscalls

l0kod opened the pull request:
Hi,

This is a WIP for Landlock (v26), a new LSM not upstream yet, but hopefully soon: https://lore.kernel.org/lkml/20201209192839.1396820-1-mic@digikod.net/
I'm working on a v27 patch series, but it should not change much the kernel part, mostly tests.

Landlock adds 3 new syscalls which are cover with this commit. I'm using syzkaller with it and it seems to work. :)

See the documentation: https://landlock.io/linux-doc/landlock-v26/userspace-api/landlock.html
and the syscall implementations: https://github.com/landlock-lsm/linux/blob/landlock-v26/security/landlock/syscall.c

Could you please give some feedback and advices to improve this patch?

I'll update this PR when Landlock will be merged upstream.

This is a team work with @vdagonneau-anssi.

dvyukov commented:
Hi Mickaël,

I am excited to see syzkaller descriptions being developed along with the new kernel subsystems!

I reviewed the descriptions themselves and they look good to me, I don't have any significant comments.
I have 2 style nits, though:
1. Suffix for the "landlock_add_rule$path_beneath" syscall:
https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions.md#names
2. Order of declarations:
https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions.md#order

When/if we are merging this, it would also be good to update syzbot configs:
https://github.com/google/syzkaller/tree/master/dashboard/config/linux
We will probably need something here:
https://github.com/google/syzkaller/blob/master/dashboard/config/linux/bits/lsm.yml
and update order of stacked LSMs for apparmour/seclinux/smack:
https://github.com/google/syzkaller/blob/master/dashboard/config/linux/bits/apparmor.yml

We can merge and start testing it once it's in linux-next. As far as I understand it's on trajectory of being merged, I think I've seen an LWN article about it or something.

And we will need signed CLA to do anything with this PR. As far as I know Microsoft has a corp CLA, you may ask me privately for details.

dvyukov commented:
>Const was initially generated with syz-extract and the "???" (because of bit shift operation) were manually fixed:

Oh, what's this? Any manual changes will be overwritten on the next regenerate.

l0kod commented:
> Hi Mickaël,

Hi Dmitry!

> 
> I am excited to see syzkaller descriptions being developed along with the new kernel subsystems!
> 
> I reviewed the descriptions themselves and they look good to me, I don't have any significant comments.
> I have 2 style nits, though:
> 
> 1. Suffix for the "landlock_add_rule$path_beneath" syscall:
>    https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions.md#names

OK, I'll rename it to landlock_add_rule$LANDLOCK_RULE_PATH_BENEATH.

> 2. Order of declarations:
>    https://github.com/google/syzkaller/blob/master/docs/syscall_descriptions.md#order

OK

> 
> When/if we are merging this, it would also be good to update syzbot configs:
> https://github.com/google/syzkaller/tree/master/dashboard/config/linux
> We will probably need something here:
> https://github.com/google/syzkaller/blob/master/dashboard/config/linux/bits/lsm.yml
> and update order of stacked LSMs for apparmour/seclinux/smack:
> https://github.com/google/syzkaller/blob/master/dashboard/config/linux/bits/apparmor.yml

OK, I'll add this with a dedicated commit.

> 
> We can merge and start testing it once it's in linux-next. As far as I understand it's on trajectory of being merged, I think I've seen an LWN article about it or something.

Yes, I think it is on a good trajectory.

> 
> And we will need signed CLA to do anything with this PR. As far as I know Microsoft has a corp CLA, you may ask me privately for details.

I'm dealing with that, it should be OK.

Thanks.

l0kod commented:
> > Const was initially generated with syz-extract and the "???" (because of bit shift operation) were manually fixed:
> 
> Oh, what's this? Any manual changes will be overwritten on the next regenerate.

I guess it's because I use the (1ULL << 0) notation: https://github.com/landlock-lsm/linux/blob/landlock-v26/include/uapi/linux/landlock.h#L114

I can take a look to fix this in syz-extract.

l0kod commented:
> (1ULL << 0) should work... I would expect.
> syz-extract does not do any manual parsing of headers, it creates a C program that prints/extracts these values. So if C compiler can chew this, it should work.

I didn't see your reply as an edit of my comment. Please add new comments to reply, it helps understand who wrote it and enables notifications.

I didn't investigate much yet, but there is already 342 matches for "???" in sys/linux/*.const, mostly for some architectures but not always: https://github.com/google/syzkaller/blob/2c1f2513486f21d26b1942ce77ffc782677fbf4e/sys/linux/dev_i915.txt.const#L3

dvyukov commented:
> I didn't see your reply as an edit of my comment. Please add new comments to reply, it helps understand who wrote it and enables notifications.

Oh, sorry, I pressed a wrong menu item.

> mostly for some architectures

Some constants are not present on some architectures, e.g. KVM is not supported on all arches.

> Could this issue be caused by incorrect include path?

This is caused by a missing include file for this const.

l0kod commented:
I fixed the `???` issue, it was because of `#include` instead of `include`. :)
I also took the styling suggestions into account.

I added two tests to help coverage, which is now 68% for security/landlock/ with syzkaller. I'm working on new ones.

One question though: do we need `breaks_returns` somewhere?

dvyukov commented:
> One question though: do we need breaks_returns somewhere?

It's only needed if the syscall makes this or subsequent syscalls return random errno values. syzkaller may use returned errno values as some coarse substitute of coverage, returning random garbage from syscalls interferes with this feature.

l0kod commented:
> It's only needed if the syscall makes this or subsequent syscalls return random errno values. syzkaller may use returned errno values as some coarse substitute of coverage, returning random garbage from syscalls interferes with this feature.

Because it's an access control system, Landlock may indirectly change the errno value of a subsequent syscall with `EACCES`. So I guess I should use `breaks_returns` for the (current) `landlock_enforce_ruleset_self()` right?

dvyukov commented:
> Because it's an access control system, Landlock may indirectly change the errno value of a subsequent syscall with EACCES. So I guess I should use breaks_returns for the (current) landlock_enforce_ruleset_self() right?

If it's only EACCES, then, no, we don't need breaks_returns. It's only for completely random integers. 1 value is not a problem and these can potentially return EACCESS on their own.

l0kod commented:
Does the errno in calls like `syscall(x, y) # EACCES` are taken into account? I didn't see any behavior change.

dvyukov commented:
> Does the errno in calls like `syscall(x, y) # EACCES` are taken into account? I didn't see any behavior change.

I don't understand the question. Please elaborate.

l0kod commented:
Comments like `# EACCES` are just cosmetic or are they used somehow by syzkaller?
e.g. https://github.com/google/syzkaller/blob/master/sys/linux/test/bpf_cgroup#L17

dvyukov commented:
> Comments like `# EPERM` are just cosmetic or are they used somehow by syzkaller?
> e.g. https://github.com/google/syzkaller/blob/master/sys/linux/test/bpf_cgroup#L17

They are not cosmetic, they are checked by pkg/runtest:
https://github.com/google/syzkaller/blob/master/pkg/runtest/run.go#L303-L319
However, note that these tests are not run automatically by anybody. You need to use tools/syz-runtest to run them manually.

l0kod commented:
The addition of fork() is redundant with #2412 but I will remove it from this branch if merged.

l0kod commented on sys/linux/test/landlock_ptrace:
```diff
@@ -0,0 +1,26 @@
+# Creates independent Landlock hierarchies.
+
+r0 = fork()
```
I guess I need to add `# requires: -arch=arm64 -arch=riscv64` because `fork()` doesn't exist for these architectures right?
Maybe something like `threaded=0` too?

dvyukov replied:
For `threaded=0` it's easier to answer.
Run:
```
make runtest
bin/syz-runtest -config manager.cfg -tests landlock_ptrace
```
And see what combinations pass/fail, and based on that you disabled what's failing (if it's expected).

dvyukov replied:
For other arches it's harder because we don't run these tests continuously on all arches. So if you add the following, I would say it's fine for now.
```
# fork is not present on these arches.
requires: -arch=arm64 -arch=riscv64
```

l0kod replied:
syz-runtest pass for all tests now!

l0kod commented:
I think this should be good now, but still not ready to merge because not in -next.

l0kod commented:
I reached 72% of coverage for security/landlock/ . The code not (yet) covered deals with race-conditions and internal errors (e.g. ENOMEM).

l0kod commented on sys/linux/test/landlock_ptrace:
```diff
@@ -0,0 +1,31 @@
+# Creates independent Landlock hierarchies.
+#
+# fork() is not present on these architectures:
+# requires: threaded=0 -arch=arm64 -arch=riscv64
```
Well… -arch=* seems to not be recognized: https://github.com/google/syzkaller/pull/2380/checks?check_run_id=1838984246#step:4:77

dvyukov replied:
The failing test does not even look at the constraints, nor it tries to execute these program, it just checks that syntax is correct:
https://github.com/google/syzkaller/blob/23a562dfb3a9986a066a1341c2cfc9e87a8fa164/pkg/csource/csource_test.go#L157-L169
As a quick unblocking fix I can only suggest to add `manual` as first constraint to match that `# requires: manual` check.

l0kod replied:
The "manual" trick doesn't work. :/

l0kod replied:
Do I need to put it in a dedicated line?

dvyukov replied:
I would assume that putting `manual` as the first constraint should work. The check checks for this substring `# requires: manual`.

l0kod commented on sys/linux/test/landlock_ptrace:
```diff
@@ -0,0 +1,32 @@
+#
+# fork() is not available for the following architectures. Set "manual" to
+# avoid test errors from csource.
+# requires: manual threaded=0 -arch=arm64 -arch=riscv64
```
I did put "manual" as the first constraint though.

l0kod replied:
The issue comes from TestParsing, not TestSysTests.

l0kod replied:
TestSysTests indeed takes the "manual" into account.

dvyukov replied:
You are right.
https://github.com/google/syzkaller/commit/2bd9619f762176527aaf28fb26e4a08b614b55df should make `arch` requirements work properly.

l0kod commented:
Is there a way to declare that a syscall changes the state of the calling thread. This could be relevant to force threaded=0 after a call to landlock_restrict_self() as well as seccomp().

dvyukov commented:
> Is there a way to declare that a syscall changes the state of the calling thread? This could be relevant to force threaded=0 after a call to landlock_restrict_self() as well as seccomp().

How would it be used by the fuzzer? We don't declare properties that we don't use (unnecessary/untested).

l0kod commented:
> > Is there a way to declare that a syscall changes the state of the calling thread? This could be relevant to force threaded=0 after a call to landlock_restrict_self() as well as seccomp().
> 
> How would it be used by the fuzzer? We don't declare properties that we don't use (unnecessary/untested).

I may have misunderstood how syzkaller works, but it seems that it can call each syscall in a dedicated thread, hence the threaded=0 option for tests. In this case, if seccomp() or landlock_restrict_self() are called on a dedicated thread, successive sibling syscalls will not be restricted.

dvyukov commented:
> > > Is there a way to declare that a syscall changes the state of the calling thread? This could be relevant to force threaded=0 after a call to landlock_restrict_self() as well as seccomp().
> > 
> > How would it be used by the fuzzer? We don't declare properties that we don't use (unnecessary/untested).
> 
> I may have misunderstood how syzkaller works, but it seems that it can call each syscall in a dedicated thread, hence the threaded=0 option for tests. In this case, if seccomp() or landlock_restrict_self() are called on a dedicated thread, successive sibling syscalls will not be restricted.

It will call subsequent syscalls on separate threads only if a previous syscall has blocked. If syscalls don't block, it will call syscalls on the same thread.

l0kod commented:
Landlock is now in linux-next: https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/tree/security/landlock?h=next-20210319&id=f00397ee41c79b6155b9b44abd0055b2c0621349

l0kod commented:
I set the initial version to 5.13 though: https://github.com/google/syzkaller/pull/2380/commits/7dd07169659d6d2eeb6457f32099085dc0bbe054#diff-708f4878ad38e69670500dbc92727f5c063a942ca2a458864a05a35657b7b907R17

dvyukov commented:
We could start testing it now, syzbot has instances on linux-next. You can set the version to `linux-next` (we later replace these with real version, once it's merged).

dvyukov commented:
We also need to regenerate kernel configs with `make configs`:
https://github.com/google/syzkaller/blob/master/dashboard/config/linux/README.md
but it requires a bunch of compilers installed, so I regenerated them to save a round trip:
https://github.com/google/syzkaller/pull/2511

The change looks good to me. Waiting for the CI to finish to merge.
Thanks for your contribution. I especially liked the added sys/linux/test/* tests.

l0kod commented:
Thanks Dmitry!

l0kod commented:
> We could start testing it now, syzbot has instances on linux-next. You can set the version to `linux-next` (we later replace these with real version, once it's merged).

I guess something is missing, this instance doesn't have Landlock yet: https://storage.googleapis.com/syzkaller/cover/ci-upstream-linux-next-kasan-gce-root.html

dvyukov commented:
linux-next is boot broken:
https://syzkaller.appspot.com/upstream
The instance is still on commit f00397ee. But there is a fixing commit for that kernel bug already, so when linux-next rebases on upstream, it should be fixed.
